### ISSUE TITLE
Boot and reactive functions?

### ISSUE THREAD

**Issue by asarkar** [bclozel/spring-boot-web-reactive#46]
How do I get all the goodies of Spring Boot while using the reactive functions?
https://spring.io/blog/2016/09/22/new-in-spring-5-functional-web-framework

**Comment by bclozel** [comment]
This experimental starter does not (yet) include the functional flavor - and maybe that flavor will need a different way of setting things up, more inline with the functional paradigm. I'll comment on this issue to keep you updated.

**Comment by sdeleuze** [comment]
@bclozel @snicoll @wilkinsona @philwebb I would like to give you my feedback coming from building [mixit](https://github.com/mix-it/mixit/) functional style Spring project (written in Kotlin but here we don't really care, it would be the same principle written in Java).

The code relevant to what is discussed in this comment are [Application.kt](https://github.com/mix-it/mixit/blob/master/src/main/kotlin/mixit/Application.kt), [Server.kt](https://github.com/mix-it/mixit/blob/master/src/main/kotlin/mixit/support/Server.kt) and [ReactorNettyServer.kt](https://github.com/mix-it/mixit/blob/master/src/main/kotlin/mixit/support/ReactorNettyServer.kt) files. Currently this application is a raw Spring Framework 5 functional API based web app, and it is quite useful to see what Spring Boot functional can bring and improve here.

A functional counterpart to `@Configuration`, `@Bean` and `@EnableXxx` could look like:

```java
public static void main(String[] args) {
	FunctionalWebApplication
		.scan("com.mypackage")
		.enableMessageSource()
		.enable(MustacheConfiguration.class)
		.enable(SpringDataMongoConfiguration.class)
		.registerBean(BarBean.class)
		.registerBean(FooWebFilter.class)
		.registerBean(FooBean.class, (context) -> new FooBean(context.getBean("barBean")))
		.registerBean(Jackson2ObjectMapperBuilder.class, () -> Jackson2ObjectMapperBuilder.locale(Locale.FRENCH))
		.runWith(ReactorNettyServer.class);
}
```

With `MustacheConfiguration` and `SpringDataMongoConfiguration` using what is currently discussed in [SPR-13779 - Programmatic Registration of Java Configuration Beans](https://jira.spring.io/browse/SPR-13779). `MustacheConfiguration` would for example registers the relevant view resolver and configurer beans for examples.

Building such application should be possible without using annotation and CGLIB proxies.

Auto-configuration based on classpath detection is not a real need I think, I would prefer controlling the behavior by enabling explicitly `FooConfiguration` classes. Generally I would expect the functional variant to give me more control about what is enabled or not, so `@Conditional` could be replaced by builder style API.

Since this configuration could be big, it would be nice to be able to compose it with different builder chains like it has been done for Spring Web Functional `RouterFunction` introduced by @poutsma.

**Comment by snicoll** [comment]
If you expect `SpringDataMongoConfiguration` to be provided by Spring Boot and if that `SpringDataMongoConfiguration` is not the regular `SpringDataMongoAutoConfiguration` (i.e. it's not  a typo), then I don't think that we should do this. If you want to take full control over your context, then you can't do that half-way. I also find that referring to the auto-configuration directly is a bad thing (auto-configuration could be internal and probably would if we didn't use them explicitly in tests).

If I misunderstood, can you please clarify where `SpringDataMongoConfiguration` would come from?

**Comment by sdeleuze** [comment]
@snicoll Good question, this point is not clear to me yet, not sure if it should reuse exactly the same `SpringDataMongoAutoConfiguration` or if it could provide an `SpringDataMongoConfiguration` more designed for the functional use case maybe using what [SPR-13779](https://jira.spring.io/browse/SPR-13779) will bring. That's just food for thoughts.

This functional flavour may require something different, but in any case there is a real need for Spring Boot to provide something, and my comment is just a first try to sketch what it could look like.

For a high level POV, I think the 4 main features I would hope to see in a Functional Spring Boot support would be:
 - Server implementations for the 5 servers we support (not provided by SPR on purpose to let Boot take care of that since configuration is not trivial)
- A functional equivalent for `@EnableXxx` and `@SpringBootApplication` annotations
- A functional builder API leveraging Spring Framework 5 `Supplier` based bean registration API
- A capacity to be composable (create an application configuration from multiple buider chains, like `RouterFunctions` and `RouterFunction` API in the Web Framework, in order to avoid having a huuuuuuuge single chain of methods for big applications

**Comment by snicoll** [comment]
> but in any case there is a real need for Spring Boot to provide something

I am not sure why. Can you expand why this has to be in Spring Boot? (re: taking full control over the context).

**Comment by asarkar** [comment]
@sdeleuze Spring Boot was born to use intelligent defaults and reduce boilerplate; otherwise you could achieve everything it does through plain vanilla Spring and explicit configurations.
[SPR-13779](https://jira.spring.io/browse/SPR-13779) refers to Spring security builders but speaking from experience, the various configurers of Spring security are quite confusing. I've ticket [SEC-3188](https://github.com/spring-projects/spring-security/issues/3393) open for more than year on that.

My point is that a bunch of configurers don't necessarily mean more control or simplified configuration. The ability to chain or compose functions is not the only thing in functional programming and going overboard with it by ditching all the auto configuration Boot provides may backfire. Even worse is a hybrid monster that has auto config and non auto config classes.

What is your primary concern? Is it that Boot is too annotation happy? Or is it that you just want Boot to step out of your way when you want complete control? I think the later is already the behavior, isn't it?

**Comment by sdeleuze** [comment]
@snicoll @asarkar Ok please forget my proposal about a more control on configuration for now, that's maybe indeed confusing to introduce a different behavior for the annotation and functional approaches. People wanting more control can just register beans themselves you are right.

Just keep in mind, and that remarks apply for both annotations based and functional approach, that RxNetty and Reactor Netty provide a single dependency for both client and server so that's maybe relevant to not rely too much on classpath detection for the Reactive support. Concretely, maybe that would be nice to enable Reactive server via a dedication `@EnableXxx` annotation rather than just classpath detection. But that's an orthogonal concern to the issue discussed here so I will create a dedicated issue for that.

So let's discuss a single and common behavior, one specified via annotation, one specified via functional API ...

Spring Boot is for me the de-facto standard way to run Spring MVC and Spring Web Reactive applications, and only supporting the annotation-based Web framework would be a big miss especially given the Java 8 focus of Spring Framework 5 (I guess that focus also applies transitively to Spring Boot 2.0). Using the functional Web + bean registration API is a perfectly valid choice for Java 8 applications, and choosing between annotation or functional API is a matter of taste. To me Spring Boot Reactive support would not complete without supporting both.

We have on purpose never integrated end-user usable server implementations in Spring Web Reactive JAR because at framework level we don't want to manage that at all. Configuring the various aspects of a server is what Spring Boot do and very well, and I think that's not a good idea to manage that on framework side. That's would not be consistent, that would be confusing. The fact that you are using the annotation-based or functional programming model should not change where we draw the line between Spring Framework and Spring Boot IMO.

But there is a real need to provide all what Spring Boot brings for functional style web application. These applications need the fat jar mechanism, need a way to enable features via an `enable(Class)` method, need a way to have a nice end-user oriented API to register beans via lambdas, etc. **I would even say that the functional web framework is just a perfect fit with Spring Boot embeddable server approach**.

In a nutshell, I would be for keeping the same Spring Boot behavior for both but allow users to create easily functional Spring Boot applications by leveraging the functional Web and bean registration Spring Framework 5 provide. When you develop a Java 8 or Kotlin application, that something which could be as successful as the annotation-based variant. I don't say one is better over the other, but both make sense and given the success of functional Web frameworks that would be a big miss for Spring Boot to not support that IMO.

**Comment by sdeleuze** [comment]
After discussing and working with @bclozel and @snicoll, I think we came to a reasonable Spring Boot initial support for the functional Web API so I think this issue could be closed when documented in the `README`(I will push a quick PR).

As demonstrated in https://github.com/mix-it/mixit/ which is now a Spring Boot application that uses the functional Web API, Spring Boot now supports scanning and taking in account all `RouterFunction` beans available in the application context. As you can see in [UserController](https://github.com/mix-it/mixit/blob/master/src/main/kotlin/mixit/controller/UserController.kt), I have chosen to keep the controller semantics while leveraging the functional Web API. I personally think this plays well, but you can also use a totally different design. Spring Boot functional supports only requires `RouterFunction` beans, they can be annotated with any kind of `@Component`.

Based on my experiments, functional bean registration is currently not a very good fit with Spring Boot in it current status so maybe something to think about later. Functional bean registration could also allow interested things used for Spring Boot internals but that's a whole different topic.
